src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingLogger.cs

```diff
@@ -0,0 +1,152 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Internal.Log;
+using static Microsoft.CodeAnalysis.LinkedFileDiffMergingSession;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal class LinkedFileDiffMergingLogger
+    {
+        private const string Id = nameof(Id);
+        private static ConcurrentDictionary<int, LogAggregator> LogAggregators = new ConcurrentDictionary<int, LogAggregator>();
+
+        internal enum MergeInfo
+        {
+            SessionsWithLinkedFiles,
+            LinkedFileGroupsProcessed,
+            IdenticalDiffs,
+            IsolatedDiffs,
+            OverlappingDistinctDiffs,
+            OverlappingDistinctDiffsWithSameSpan,
+            OverlappingDistinctDiffsWithSameSpanAndSubstringRelation,
+            InsertedMergeConflictComments,
+            InsertedMergeConflictCommentsAtAdjustedLocation
+        }
+
+        internal static void LogSession(Workspace workspace, LinkedFileDiffMergingSessionInfo sessionInfo)
+        {
+            if (sessionInfo.LinkedFileGroups.Count > 1)
+            {
+                int correlationId = GetCorrelationId(workspace);
+
+                LogNewSessionWithLinkedFiles(correlationId);
+                LogNumberOfLinkedFileGroupsProcessed(correlationId, sessionInfo.LinkedFileGroups.Count);
+
+                foreach (var groupInfo in sessionInfo.LinkedFileGroups)
+                {
+                    LogNumberOfIdenticalDiffs(correlationId, groupInfo.IdenticalDiffs);
+                    LogNumberOfIsolatedDiffs(correlationId, groupInfo.IsolatedDiffs);
+                    LogNumberOfOverlappingDistinctDiffs(correlationId, groupInfo.OverlappingDistinctDiffs);
+                    LogNumberOfOverlappingDistinctDiffsWithSameSpan(correlationId, groupInfo.OverlappingDistinctDiffsWithSameSpan);
+                    LogNumberOfOverlappingDistinctDiffsWithSameSpanAndSubstringRelation(correlationId, groupInfo.OverlappingDistinctDiffsWithSameSpanAndSubstringRelation);
+                    LogNumberOfInsertedMergeConflictComments(correlationId, groupInfo.InsertedMergeConflictComments);
+                    LogNumberOfInsertedMergeConflictCommentsAtAdjustedLocation(correlationId, groupInfo.InsertedMergeConflictCommentsAtAdjustedLocation);
+
+                    if (groupInfo.InsertedMergeConflictComments > 0 ||
+                        groupInfo.InsertedMergeConflictCommentsAtAdjustedLocation > 0)
+                    {
+                        Logger.Log(FunctionId.Workspace_Solution_LinkedFileDiffMergingSession_LinkedFileGroup, SessionLogMessage.Create(correlationId, groupInfo));
+                    }
+                }
+            }
+        }
+
+        private static int GetCorrelationId(Workspace workspace)
+        {
+            int correlationId = workspace.GetHashCode();
```

Comment:
Talked to @jmarolf offline, we think it'll be better to keep track per VS session instead of try to get this granular. Can we use a more global session id to fill this correlationId for the entire session?